src/PostgREST/Query/SqlFragment.hs

```diff
@@ -334,3 +335,12 @@ unknownLiteral = unknownEncoder . encodeUtf8
 intercalateSnippet :: ByteString -> [SQL.Snippet] -> SQL.Snippet
 intercalateSnippet _ [] = mempty
 intercalateSnippet frag snippets = foldr1 (\a b -> a <> SQL.sql frag <> b) snippets
+
+-- the "ctid" system column is always available to tables
+mutRangeF :: QualifiedIdentifier -> [FieldName] -> (SqlFragment, SqlFragment)
+mutRangeF mainQi rangeId = (
+    BS.intercalate " AND " $
+          (\col -> pgFmtColumn mainQi col <> " = " <> pgFmtColumn (QualifiedIdentifier mempty "pgrst_affected_rows") col) <$>
+          (if null rangeId then ["ctid"] else rangeId)
+  , if null rangeId then pgFmtColumn mainQi "ctid" else BS.intercalate ", " (pgFmtColumn mainQi <$> rangeId)
+  )
```

Comment:
Given what [I mentioned below](https://github.com/PostgREST/postgrest/pull/2195#issuecomment-1079736464), it does make a lot of sense to order by `ctid` only and not by PK, because doing this:

```sql
INSERT INTO projects VALUES (1, 'Windows 7', 1);
INSERT INTO projects VALUES (2, 'Windows 10', 1);
INSERT INTO projects VALUES (5, 'Orphan', NULL);
INSERT INTO projects VALUES (4, 'OSX', 2);
INSERT INTO projects VALUES (3, 'IOS', 2);
```

Gives through http:

```http
GET /projects

[{"id":1,"name":"Windows 7","client_id":1},
 {"id":2,"name":"Windows 10","client_id":1},
 {"id":5,"name":"Orphan","client_id":null},
 {"id":4,"name":"OSX","client_id":2},
 {"id":3,"name":"IOS","client_id":2}]
```

Which is the same as doing `GET /projects?order=ctid` but not the same as doing `GET /projects?order=id`.

Reply:
So basically with `ctid` we can ensure that `DELETE /projects?limit=2` deletes the same rows as the ones obtained through `GET /projects?limit=2`.

Reply:
The only thing bad about preferring `ctid` is that you have to give explicit SELECT privilege to the column for it work(or a full wide SELECT) and that is weird - giving the PK columns a SELECT privilege is a more usual operation.

Reply:
So because of the above - the need to do an special `GRANT SELECT(ctid)` for `limit` to work - I think in fact `ctid` should be left out and `limit` should depend on PKs.

Later on, it would also be weird to tell users to expose `ctid` in ther VIEWs for `limit` to work.

Reply:
> The only thing bad about preferring ctid is that you have to give explicit SELECT privilege to the column for it to work(or a full wide SELECT) and that is weird - giving the PK columns a SELECT privilege is a more usual operation.

I don't think that's weird.

It should be a pretty descriptive error message anyway (lacking permission to select ctid) and we can add a hint about that on the feature page.

The advantage of using ctid:
- Consistent ordering with `GET` for both views with pks and without.
- Easier to explain ordering in the docs, if we do at all.
- More flexible, because you might want to explicitly **not** give `SELECT` on (a part of) the primary key. `ctid` is surely not sensitive, so granting a select for those bulk actions on it shouldn't be a problem.

Reply:
> ctid is surely not sensitive

Yeah, this is a nice advantage.

> Later on, it would also be weird to tell users to expose ctid in ther VIEWs for limit to work.

What about the above though? For views, it would be invasive to tell the user to explicitly expose the `ctid`. Usually views are used for exposing just the columns needed for the frontend.

Or perhaps we can have an additional rule, that for views you have to explicitly `order` by a column for the limited `update/delete` to work?

Reply:
Actually... it just occured to me, that we should just not use `ctid` for ordering at all. Without `?order=` this will then be exactly the same as `GET` - we don't use `ctid` there either. And PG sort order is not guaranteed, so we don't know whether it will always match up with `ctid`.

We can use `ctid` for the CTE join right now. And we can extend that later to views with whatever columns we use for that. But we don't need either for ordering.

Reply:
> And PG sort order is not guaranteed, so we don't know whether it will always match up with ctid.

True. And I also like the idea of not doing an implicit `order`. So using the `ctid` as default for both tables/views it is.

> We can use ctid for the CTE join right now. And we can extend that later to views with whatever columns we use for that.

How about if for both tables and views we use the explicit `order` column(s) for the join, that way is up to the client to decide the criteria. This would remove the need for smartness from the internal cache and also provide a way out of requiring `ctid` on views.

Reply:
This is quite an interesting idea. I'm not sure about whether I like the following suggestion, but if we take it a step further, we could do the following:
- Disallow any mutation request with `?limit` but without `?order`.
- Always use the `?order` columns for both the "inside the CTE `ORDER BY`" **and** the join condition

The great benefit at this point would be: This would support both tables and views without a problem.

The downside: To work properly, this would require the user to always use columns in `order` that form a unique identifier for a single row. Otherwise the query could mutate more rows than requested. As a first step, we could warn very explicitly about this behavior in the docs and tell the user that they need to use unique columns.

In a later step we could add an implicit ["required row count filter"](https://github.com/PostgREST/postgrest/pull/2164) to force a hard-limit on the mutated rows using something like `!row_count=lt.<limit>`. This would basically throw an error if the user used some non-unique set of columns, which would lead to a delete bigger than the requested limit. We should certainly not implement that second step right now, but wait for #2164 - but if we end up with this, this could be a very nice solution to support the whole thing on views properly - without any of those "need to expose column XYZ" challenges. It would be up to the user to know which columns fulfill the uniqueness requirement.

Reply:
> > The downside: To work properly, this would require the user to always use columns in order that form a unique identifier for a single row.
> 
> I think I can a put a nice error and even suggest trying the pks available on the schema cache, that should help with DX.

You can only show this error message in the case of "`?limit` without `?order` specified". When a user chooses to `order` on non-pk columns, you can't show this error, because you'd need to detect the uniqueness of those columns again - which is exactly what we have trouble with for views. The answer to that is my suggestion below...
 
> > In a later step we could add an implicit "required row count filter" to force a hard-limit on the mutated rows using something like !row_count=lte.
> 
> So like a `db-max-rows` for the `row_count` right? That's something I've been wanting as well. Unlike `db-max-rows` it would make sense to override it client side so if an operation requires mutating a large amount of rows the client must be explicit about that.

Now, I don't really understand what you mean by "`db-max-rows` for `row_count`, but I have a feeling there is mis-understanding here.

What I am thinking is the following:

Given this schema/data:

```sql
create view v as select pk, 'static' AS static from some_table;
-- insert 3 rows into some_table
```

Those 3 rows would have different PKs, but the same `static` column.

A query like this would work fine:

```http
DELETE /v?limit=2&order=pk
```

It would delete exactly 2 rows all the time, because `pk` is unique.

However, a query like this would not work as expected:

```http
DELETE /v?limit=2&order=static
```

Since `static` is the same for all 3 rows, it would select two of those rows in the CTE - but then join on that column and actually `DELETE` all 3 rows.

However, if we instead run this query implicitly as the following:
```http
DELETE /v!row_count=lte.2?limit=2&order=static
```

Then we would receive an error, because the actual mutation included 3 rows, which is bigger than the limit requested. In this case the transaction would be rolled back anyway.

If we can catch this error properly, and then show a better hint like "you need to use unique columns in `order`" - great.

Is this what you understood, too?

Reply:
I see. A server-side default, which is overridable from the client-side, would be completely unrelated to this issue here, though. The implicit row-count here would just be to avoid mistakes in the `order` clause - while the default you are proposing would be a protection against unlimited delete.

Reply:
> However, if we instead run this query implicitly as the following:
> DELETE /v!row_count=lte.2?limit=2&order=static

@wolfgangwalther Now that I've realized, this would mean that `limit` needs `row_count` and that for this feature both might as well be the same.

So how about this. I'll go ahead and apply the `row_count` logic for the `limit` but will not expose the actual `row_count` special filter. We can add that one later on when we settle on a syntax.

(I do see `row_count` being useful in other ways but for this feature we can apply it implicitly as you mentioned)

Reply:
And now that I think about it. If we enforce `row_count` implicitly on `limit`, then it would be safe to assume the PK cols for the `order/join` on views as well?

So the [above problem](https://github.com/PostgREST/postgrest/pull/2195#issuecomment-1079249217) would be no more?

Reply:
> So how about this. I'll go ahead and apply the `row_count` logic for the `limit` but will not expose the actual `row_count` special filter.

Ok.

> I mean because we'd be already fulfilling our part of the contract - the affected rows would not exceed the `limit`. And if the VIEW query reduces the amount of rows mapped to the PKs this would not be an issue.

I think it would still have very odd interactions with `offset`. For me, it'd be much simpler to implement and explain to always require and use the columns in `order` - and hint in the docs and error messages at the requirements for that. We'd need that for views without inferred PK columns anyway - and this way we don't have a complicated mix of what to do in which situation.